### DISCUSSION TITLE
Local variables in cc65

### DISCUSSION THREAD

**Discussion by amukh1** [cc65/cc65#2430] [Q&A]
So, I have been having some trouble with local variables and the stack and function parameters in cc65. To start,

```c
int main() {
    char x = 4;
    char y = 5;

    return 0;
}
```
This seems to allocate both x and y at 0x00, (overriding x) instead of using the stack operations, why is this? Is this supposed to happen?

Here is my cfg:
```cfg
MEMORY {
    ZP:     start = $0002, size = $00FE, type = rw, define = yes;
    RAM:    start = $0200, size = $0FFF, type = rw, define = yes;
    ROM:    start = $8000, size = $8000, type = ro, define = yes;
    STACK: start = $1000, size = $0FFF, type = rw, define = yes;
}

SEGMENTS {
    ZEROPAGE: load = ZP,  type = zp;
    DATA:     load = RAM, type = rw;
    CODE:     load = ROM, type = ro;
    RODATA:  load = ROM, type = ro;
    BSS:      load = RAM, type = bss, define = yes;
    STACK:    load = STACK, type = bss, optional = yes;
    ONCE:     load = ROM, type = ro, optional = yes;
}

SYMBOLS { 
    __STACKSIZE__: type = weak, value = $0200; 
}```
I thought it might be optimization so I tried using both variables by passing them into other functions but that doesn't seem to fix it either.

And that leads into the next problem, parameters are also behaving.. funky
```c
void test(char a, char b) {
    *(char*)0x10 = a;
    *(char*)0x11 = b;
}

int main() {
    test(0x12, 0x34);

    return 0;
}
```
That (as before) just leaves the second parameter at 0x00 by overwriting the first one, and then neither is accessible in the body.
![image](https://github.com/cc65/cc65/assets/85003861/b4012e01-d4a3-4995-b57e-817805425247)
(Hexdumped end-state memory)
I figured I wouldn't be able to see anything since they are local but I ended up seeing these side effects. I am sorry if these are trivial questions, I am new to the 6502 😅

**Comment by mrdudz** [comment]
What target is this? Where the stack is allocated entirely depends on the startup code of the target. Please give a complete example, ie include the commandline used to compile and link the above program.

**Reply by amukh1** [reply]
This is what I'm using to compile test.c:
```sh
cc65 -Oirs test.c -o test.s    
ca65  test.s -o test.o
ca65 startup.s -o startup.o
ld65 -C test.cfg startup.o test.o -o test c64.lib
```


test.c:
```c
#include <stdlib.h>
#include <stdint.h>
#include "./include/6502.h"

#pragma bss-name(push, "BSS")

void test(char a, char b) {
    *(char*)0x10 = a;
    *(char*)0x11 = b;
}

int main() {
    char x = 4;
    char y = 5;

    return 0;
}
```

startup.s:
```asm
    .export __STARTUP__
    .import _main
    .importzp sp

__STARTUP__:
    ; Initialize the stack pointer
    ldx #$ff
    txs

    ; Call the main function
    jsr _main

    ; Endless loop after main returns
    ; jmp *

    ; Interrupt vectors
    .word $0000 ; NMI
    .word __STARTUP__ ; RESET
    .word $0000 ; IRQ
```

This example allocates both x and y to 0x00 and i'm not sure what exactly is going wrong.
This is the output from cc65 of test.c:

```asm
;
; File generated by cc65 v 2.18 - Ubuntu 2.18-1
;
	.fopt		compiler,"cc65 v 2.18 - Ubuntu 2.18-1"
	.setcpu		"6502"
	.smart		on
	.autoimport	on
	.case		on
	.debuginfo	off
	.importzp	sp, sreg, regsave, regbank
	.importzp	tmp1, tmp2, tmp3, tmp4, ptr1, ptr2, ptr3, ptr4
	.macpack	longbranch
	.forceimport	__STARTUP__
	.export		_test
	.export		_main

; ---------------------------------------------------------------
; void __near__ test (unsigned char, unsigned char)
; ---------------------------------------------------------------

.segment	"CODE"

.proc	_test: near

.segment	"CODE"

	jsr     pusha
	ldy     #$01
	lda     (sp),y
	sta     $0010
	dey
	lda     (sp),y
	sta     $0011
	jmp     incsp2

.endproc

; ---------------------------------------------------------------
; int __near__ main (void)
; ---------------------------------------------------------------

.segment	"CODE"

.proc	_main: near

.segment	"CODE"

	lda     #$04
	jsr     pusha
	lda     #$05
	jsr     pusha
	ldx     #$00
	txa
	jmp     incsp2

.endproc
```

If I replace the ``jsr pusha`` lines with PHA it seems to work, but cc65 is using pusha

**Comment by mrdudz** [comment]
Well, you are basically compiling for the C64 target, and then you don't use the proper startup code.

The stack that is used for parameter passing is NOT the 6502 stack for that matter - its a software stack (that usually lives at the end of memory).

You want to look at the crt0.s of the C64 target to see how to set up things correctly

**Reply by amukh1** [reply]
Is there a way to compile not for the C64? ;-; I tried removing it but the output keep referencing pusha, sp, etc. Do I need to write my own?

**Comment by mrdudz** [comment]
Those functions are not C64 specific - thats how cc65 works, you will need to initialize the software stack to match your target (what target is it?)

That said, you can use -t none to get a generic setup

**Reply by amukh1** [reply]
I don't really have a specific target, I am trying to run the programs on an emulator I am building. Everything _should_ be working but I keep running into problems. I don't really want to compile to a specific machine I just want to run it on my plain emulator xD

This is the main thing that is confusing me and if I get past this I think the parameter issues will solve themsleves,
```c
int main() {
    char x = 4;
    char y = 5;
    return 0;
}
```
(same cfg) 
is allocating x and y at the same place
```asm
.proc	_main: near

.segment	"CODE"

	lda     #$04
	jsr     pusha
	lda     #$05
	jsr     pusha
	ldx     #$00
	lda     #$00
	jmp     L000A
L000A:	jsr     incsp2
	rts
```
this is the corresponding code out of cc65, shouldn't the pusha subroutine be incrementing/decrementing the sp to get a different address for x and y? 

If I replace my cfg with ``-t none`` (keeping startup.o)
It doesn't allocate anything at 0x00 and randomly allocates 0x05 and 0x80 near 0xf0
![image](https://github.com/cc65/cc65/assets/85003861/681cbcf7-ab2f-41af-9077-f5227123f456)

And if I remove startup.o, well idk what I'm looking at
![image](https://github.com/cc65/cc65/assets/85003861/13f3bf79-08f7-4baf-a39b-2813be2365c7)
I tried looking in crt0.s but there wasn't anything about pusha

**Reply by Marrin** [reply]
Yes `pusha` should be decrementing `sp`.  And it does: https://github.com/cc65/cc65/blob/master/libsrc/runtime/pusha.s

**Comment by mrdudz** [comment]
Again: The variables are allocated on the software stack. If you don't initialize the software stack(pointer), then it may point to 0, or whatever else.

https://github.com/cc65/cc65/blob/master/libsrc/none/crt0.s

look what it does, this is the absolute minimum you'll have to do.
